### PR TITLE
Convert between SBOM formats

### PR DESCRIPTION
**Pull request by jonasagx**
closes https://github.com/anchore/syft/issues/563

## To-Do
- [x]  Move convert command to new command design 
- [x] Figure out why `TestEncodeDecodeEncodeCycleComparison` fails for SPDX: spdx's `ToSyftModel` currently doesn't care about metadata, let's make it care
- [x] Unit test new code
- [x] Integration test new command
- [ ]  <del>Move SPDX encoder to SPDX's go lib and add tests</del> it doesn't seem necessary 
- [ ] <del> Refactor encode-decode-encode-again tests for SPDX</del>
- [ ] <del>Compare data loss between formats</del>

### CODE REVIEW

**Comment by jonasagx** [comment]
Got stuck on moving `convert` to  new command structure. The party bus subscription keeps on hanging.

**Comment by jonasagx** [comment]
Between conversions, from SPDX-json to SPDX-json, a couple fields come through with different values or cases. The problematic fields are:

- name
- documentNamespace
- sourceInfo
- licenseDeclared
- licenseConcluded
- originator

**Comment by jonasagx** [comment]
We currently don't test SPDX validation, but we have code for that. I will add it so I have a touchstone to validate conversion results:

```
 --- FAIL: TestSPDXJSONDecoder/bad/example7-bin.spdx.json (0.00s)
        decoder_test.go:67:
                Error Trace:    decoder_test.go:67
                Error:          Received unexpected error:
                                unable to decode spdx-json: no colon found although DocumentRef- prefix present
                Test:           TestSPDXJSONDecoder/bad/example7-bin.spdx.json
    --- FAIL: TestSPDXJSONDecoder/bad/example7-go-module.spdx.json (0.00s)
```

File: internal/formats/common/spdxhelpers/to_syft_model.go [hunk 1]
```diff
@@ -42,6 +49,35 @@ func ToSyftModel(doc *spdx.Document2_2) (*sbom.SBOM, error) {
 	return s, nil
 }
 
+// NOTE(jonas): this is a hack, SPDX doesn't seem to support a way
+// to describe what the SBOM is about: image, directory for example,
+// so I am trying to parse it from namespace. Syft generated SBOMs
+// have a type encoded in this field, so this approach works well
+// for SBOMs generated by Syft.
+func extractSourceFromNamespace(ns string) (src source.Metadata) {
```
Comment (jonasagx) [inline internal/formats/common/spdxhelpers/to_syft_model.go:57]: This func was added to move us closer to a functional decode/encode/decode cycle for SPDX. Without this info source.Metadata won't have a scheme and cannot be decoded.

File: README.md [hunk 2]
```diff
@@ -25,6 +25,7 @@ A CLI tool and Go library for generating a Software Bill of Materials (SBOM) fro
 - Linux distribution identification
 - Works seamlessly with [Grype](https://github.com/anchore/grype) (a fast, modern vulnerability scanner)
 - Able to create signed SBOM attestations using the [in-toto specification](https://github.com/in-toto/attestation/blob/main/spec/README.md)
+- Convert between SBOM formats, such as CycloneDX (JSON, XML), SPDX, and Syft's own format.
```
Comment (luhring) [inline README.md:28]: I think we can leave out the "(JSON, XML)" qualifier — there are multiple encoding formats for SPDX too, but we don't need to list them here in this summary IMHO

File: README.md [hunk 3]
```diff
@@ -92,6 +93,24 @@ The above output includes only software that is visible in the container (i.e.,
 syft <image> --scope all-layers
 ```
 
+#### Format conversion - __experimental__
+
+Syft can convert between different SBOM formats, such as CycloneDX, SPDX, and Syft's own format. The ability to convert SBOMs helps saving the time and effort it would take to recreate SBOMs from container images, or file systems.
+
+``` bash
+syft convert sbom.syft.json -o cyclonedx-json=sbom.cdx.json
```
Comment (luhring) [inline README.md:101]: I would generalize this introduction of the command, like you'd see in a "synopsis" view of a CLI command — instead of starting with a specific example. Maybe something like:

```
syft convert <ORIGINAL-SBOM-FILE> -o <NEW-SBOM-FORMAT>[=<NEW-SBOM-FILE>]
```

And then maybe include an example at the end of this "format conversion" section

File: README.md [hunk 4]
```diff
@@ -92,6 +93,24 @@ The above output includes only software that is visible in the container (i.e.,
...
+```
+
+This feature is experimental and data might be lost when converting formats. Packages are the main SBOM component easily transferable across formats, while files and relationships are more likely to lose data. 
+
+The supported formats are:
```
Comment (luhring) [inline README.md:106]: We may have covered this, but refresh me: do we not want to support all formats?

File: README.md [hunk 5]
```diff
@@ -92,6 +93,24 @@ The above output includes only software that is visible in the container (i.e.,
 syft <image> --scope all-layers
 ```
 
+#### Format conversion - __experimental__
+
+Syft can convert between different SBOM formats, such as CycloneDX, SPDX, and Syft's own format. The ability to convert SBOMs helps saving the time and effort it would take to recreate SBOMs from container images, or file systems.
```
Comment (luhring) [inline README.md:98]: >The ability to convert SBOMs helps saving the time and effort it would take to recreate SBOMs from container images, or file systems.

Suggestion:

The ability to convert existing SBOMs means you can create SBOMs in different formats quickly, without the need to regenerate the SBOM from scratch, which may take significantly more time.

File: cmd/syft/cli/convert.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	convertExample = `  {{.appName}} {{.command}} img.syft.json -o spdx-json                      convert a syft SBOM to spdx-json, output goes to stdout in table format, by default
+  {{.appName}} {{.command}} img.syft.json -o cyclonedx-json=img.cdx.json    convert a syft SBOM to CycloneDX, output goes to a file named img.cdx.json
+`
+)
+
+func Convert(v *viper.Viper, app *config.Application, ro *options.RootOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "convert sbom.json -o [FORMAT]",
+		Short: "Convert between SBOM formats",
+		Long:  "[Experimental] Convert SBOM files to, and from, SPDX, CycloneDX and Syft's format. For more info about data loss between formats read Syft's README on https://github.com/anchore/syft",
```
Comment (luhring) [inline cmd/syft/cli/convert.go:24]: >read Syft's README on https://github.com/anchore/syft

I think we can just say "see [insert direct link to the format conversion section]"

File: internal/formats/spdx22json/decoder_test.go [hunk 7]
```diff
@@ -61,6 +63,12 @@ func TestSPDXJSONDecoder(t *testing.T) {
 			f, err := os.Open("test-fixtures/spdx/" + test.path)
 			assert.NoError(t, err)
 
+			if !test.fail {
+				err = Format().Validate(f)
+				require.NoError(t, err)
+				_, err = f.Seek(0, io.SeekStart)
+			}
+
```
Comment (luhring) [inline internal/formats/spdx22json/decoder_test.go:71]: What is the intent of this block?

File: internal/formats/spdx22json/to_format_model.go [hunk 8]
```diff
@@ -153,6 +153,7 @@ func toFiles(s sbom.SBOM) []model.File {
 				},
 				// required, no attempt made to determine license information
 				LicenseConcluded: "NOASSERTION",
+				// LicenseConcluded: "NONE", 19 changes
```
Comment (luhring) [inline internal/formats/spdx22json/to_format_model.go:156]: Why was this inserted?

File: test/integration/convert_test.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/anchore/syft/internal/formats/syftjson"
+	"github.com/anchore/syft/internal/formats/table"
+	"github.com/anchore/syft/syft"
+	"github.com/anchore/syft/syft/source"
+	"github.com/stretchr/testify/require"
+)
+
+// TestConvertCmd tests if the converted SBOM is a valid document according
+// to spec.
+// FIXME: This test can, but currently does not, check the converted SBOM content. It
```
Comment (luhring) [inline test/integration/convert_test.go:20]: nit: I'd say "TODO" instead of "FIXME", so that we don't need to search for an additional string as we find TODO items

File: test/integration/encode_decode_cycle_test.go [hunk 10]
```diff
@@ -99,3 +107,25 @@ func TestEncodeDecodeEncodeCycleComparison(t *testing.T) {
...
+}
+
+// func saveTempFile(t *testing.T, b []byte, prefix string) {
+// 	f, err := ioutil.TempFile(os.TempDir(), prefix)
+// 	assert.NoError(t, err)
+// 	s, err := f.Write(b)
+// 	assert.NoError(t, err)
+// 	t.Logf("you've got %d bytes written to %s", s, f.Name())
+// 	f.Close()
+// }
```
Comment (luhring) [inline test/integration/encode_decode_cycle_test.go:131]: Do we need to check this in?

File: cmd/syft/cli/commands.go [hunk 11]
```diff
@@ -75,6 +77,11 @@ func New() (*cobra.Command, error) {
 	if err != nil {
 		return nil, err
 	}
+	// convert also uses flags from the packagesCmd since it generates an sbom
+	err = po.AddFlags(convertCmd, v)
```
Comment (spiffcs) [inline cmd/syft/cli/commands.go:81]: There's not a lot in the PR description so apologies if I missed this in the linked issue, but why should `covert` be able to generate an SBOM? 

```
syft convert sbom.syft.json -o cyclonedx-json=sbom.cdx.json
```

I don't see any of the common packages flags here and lean towards keeping convert pure as it's own command with it's own flags atm.

File: cmd/syft/cli/convert.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+				return fmt.Errorf("invalid application config: %w", err)
+			}
+			newLogWrapper(app)
+			logApplicationConfig(app)
+			return validateArgs(cmd, args)
+		},
+		SilenceUsage:  true,
+		SilenceErrors: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			newLogWrapper(app)
```
Comment (spiffcs) [inline cmd/syft/cli/convert.go:40]: I don't think we don't need to call newLogWrapper twice here

File: internal/formats/common/spdxhelpers/to_syft_model.go [hunk 13]
```diff
@@ -24,7 +25,13 @@ func ToSyftModel(doc *spdx.Document2_2) (*sbom.SBOM, error) {
 
 	spdxIDMap := make(map[string]interface{})
 
+	src := source.Metadata{}
+	if doc.CreationInfo != nil {
+		src = extractSourceFromNamespace(doc.CreationInfo.DocumentNamespace)
+	}
+
 	s := &sbom.SBOM{
+		Source: src,
```
Comment (spiffcs) [inline internal/formats/common/spdxhelpers/to_syft_model.go:34]: Is this a required field in the struct? Should we only set this field `if doc.CreationInfo != nil`

File: test/integration/convert_test.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+// might be useful to do that in the future, once we gather a better understanding of
+// what users spect from the convert command.
```
Comment (spiffcs) [inline test/integration/convert_test.go:22]: spect => expect

File: test/integration/convert_test.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+			err = convert.Run(ctx, app, []string{f.Name()})
+			require.NoError(t, err)
+			stdw.Close()
+
+			out, err := ioutil.ReadAll(stdr)
+			require.NoError(t, err)
+
+			os.Stdout = originalStdout
+			// t.Logf("out: %s", out)
```
Comment (spiffcs) [inline test/integration/convert_test.go:56]: Do you think we should keep the commented-out code here? I usually err on the side of making a call to either remove commented code or enable it in a PR. Having it build-up is usually a code smell.

File: test/integration/encode_decode_cycle_test.go [hunk 16]
```diff
@@ -35,6 +35,16 @@ func TestEncodeDecodeEncodeCycleComparison(t *testing.T) {
 		redactor     func(in []byte) []byte
 		json         bool
 	}{
+		// TODO: SPDX's lib and our current implementation of the encoder function creates mismatches with metadata fields.
```
Comment (spiffcs) [inline test/integration/encode_decode_cycle_test.go:38]: Same question re: commented test/commented code from above

**Review by spiffcs — COMMENTED** [review]
Looks good so far! 

Comments were added about the implementation of the flags to see if we can somehow make it so that this command does not entangle itself with the package command flags. 

Also, minor `NIT` comments about double function calls, comments, and copy-pasted functionality across the command package.

File: README.md (hunk 4 above)
Comment (jonasagx) [reply README.md:106]: We support formats with wide community usage AND encode/decode support. The recent GitHub format is niche and doesn't have a decoder, therefore was excluded.

The exception to the rule above is table format, which is the default output for two reasons:
1. it matches syft's default SBOM output
2. it is handy, as a summary, to print SBOMs as tables

Comment (jonasagx) [reply README.md:106]: @luhring I will add the explanation above to the README so others can understand the design decision.

File: internal/formats/spdx22json/decoder_test.go (hunk 7 above)
Comment (jonasagx) [reply internal/formats/spdx22json/decoder_test.go:71]: The intent is to use SPDX's lib to validate our home-brew encoder, no changes to our encoder were made but it is good to have fallback tests for the future.

File: internal/formats/spdx22json/to_format_model.go (hunk 8 above)
Comment (jonasagx) [reply internal/formats/spdx22json/to_format_model.go:156]: ops, it's a leftover from output comparison, clean it up

File: cmd/syft/cli/commands.go (hunk 11 above)
Comment (jonasagx) [reply cmd/syft/cli/commands.go:81]: You're right, we can keep it isolated flag-wise.

File: internal/formats/common/spdxhelpers/to_syft_model.go (hunk 13 above)
Comment (jonasagx) [reply internal/formats/common/spdxhelpers/to_syft_model.go:34]: This is best effort to define `source.Scheme`, without `source.Scheme` we can't finish the conversion cycle.

File: test/integration/encode_decode_cycle_test.go (hunk 16 above)
Comment (jonasagx) [reply test/integration/encode_decode_cycle_test.go]: removed 👍🏼

**Comment by jonasagx** [comment]
@spiffcs yeah, I removed the AddFlags part from `cli/command.go` 
```diff
-	// convert also uses flags from the packagesCmd since it generates an sbom
-	err = po.AddFlags(convertCmd, v)
-	if err != nil {
-		return nil, err
-	}
```

On code organization: sure, I will make convertCmd near the other cmds.

File: README.md [hunk 17]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
...
+
+We support formats with wide community usage AND good encode/decode support by Syft. The recent GitHub format is niche and doesn't have a decoder, therefore was excluded.
+
+The exception to the rule above is table format, which is the default output for two reasons:
+1. it matches syft's default SBOM output
+2. it is handy, as a summary, to print SBOMs as tables
+
+Conversion example:
+```bash
+syft alpine:latest -o json=sbom.syft.json
```
Comment (spiffcs) [inline README.md:122]: I'm a little confused and sorry I didn't catch this the first time. Why is convert aliased over the root command and why has the -o flag changed here from supplying a format to this new syntax of `json=sbom.syft.json`?

File: cmd/syft/cli/convert/convert.go [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+	for _, o := range outputOptions {
+		log.Debugf("writer options: %+v", o.Format)
+	}
+
+	return writer, nil
+}
+
+// parseOptions utility to parse command-line option strings and retain the existing behavior of default format and file
+func parseOptions(outputs []string, defaultFile string) (out []sbom.WriterOption, formats []sbom.Format, errs error) {
```
[the remark this answers is not in the document]
Comment (spiffcs) [reply cmd/syft/cli/convert/convert.go:97]: Awesome! Makes sense why they are a little different - is there a way we can update it so that we don't copy and paste over with those small changes? Maybe allow writer to accept which are valid/supported formts?

cc @anchore/tools for additional comments if I'm totally off on this suggestion.

File: internal/formats/spdx22json/decoder_test.go [hunk 19]
```diff
@@ -61,6 +63,12 @@ func TestSPDXJSONDecoder(t *testing.T) {
 			f, err := os.Open("test-fixtures/spdx/" + test.path)
 			assert.NoError(t, err)
 
+			if !test.fail {
+				err = Format().Validate(f)
+				require.NoError(t, err)
+				_, err = f.Seek(0, io.SeekStart)
```
Comment (spiffcs) [inline internal/formats/spdx22json/decoder_test.go:69]: Do we need to reset the offset here because Validate manipulates the file?

File: README.md [hunk 20]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
 syft <image> --scope all-layers
 ```
 
+#### Format conversion - __experimental__
+
+The ability to convert existing SBOMs means you can create SBOMs in different formats quickly, without the need to regenerate the SBOM from scratch, which may take significantly more time.
+
+``` bash
```
Comment (luhring) [inline README.md:100]: nit: I just noticed the `bash` interpreter is making the synopsis syntax look strange since it's not real bash. Maybe we just use `text` or something here.

File: README.md [hunk 21]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
 syft <image> --scope all-layers
 ```
 
+#### Format conversion - __experimental__
```
Comment (luhring) [inline README.md:96]: nit: I think the `__` use here has no effect, it's bolding what's already bold due to being a header.

Also, what's the thinking behind using H4 here?

Suggestion: `### Format conversion (experimental)`

File: internal/formats/spdx22json/decoder_test.go (hunk 7 above)
Comment (luhring) [reply internal/formats/spdx22json/decoder_test.go:71]: I don't understand yet... How does this fit in with the conversion functionality? And what is this guarding us from?

**Review by spiffcs — COMMENTED** [review]
If you have time maybe we can get @luhring on a quick call to talk about the example and updated flags. I'm so sorry I missed this before, but I'm a little confused as to how `convert` can be aliased to the root command with a newly formatted `-o` flag.

File: test/integration/encode_decode_cycle_test.go [hunk 22]
```diff
@@ -69,15 +67,13 @@ func TestEncodeDecodeEncodeCycleComparison(t *testing.T) {
 				format := syft.FormatByID(test.formatOption)
 				require.NotNil(t, format)
 
-				by1, err := syft.Encode(originalSBOM, format)
-				assert.NoError(t, err)
+				by1 := encode(t, originalSBOM, format)
```
Comment (spiffcs) [inline test/integration/encode_decode_cycle_test.go:70]: I'm also a little confused about the diff here. I don't see changes to `syft.Encode` and thought that this test was specifically for checking if changes to that method cause lossy behavior across conversions. Why have we updated this to a custom test function that now uses the `format.Encode` method?

File: README.md (hunk 17 above)
Comment (jonasagx) [reply README.md:122]: There is no conversion happening here. What I should have explained better is in this example: first generate an SBOM with syft's format, then convert it.

Comment (jonasagx) [reply README.md:122]: Conversion happens right below, line 123

File: internal/formats/spdx22json/decoder_test.go (hunk 7 above)
Comment (jonasagx) [reply internal/formats/spdx22json/decoder_test.go:71]: This is making sure future changes to our SPDX encoder will follow their (SPDX) official specs.

Comment (jonasagx) [reply internal/formats/spdx22json/decoder_test.go:71]: Actually, I think a better place for it is with encode/decode tests.

File: cmd/syft/cli/convert/convert.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (jonasagx) [reply cmd/syft/cli/convert/convert.go]: Good point, I organized it the function can be used by both package and convert cmds.

File: README.md (hunk 21 above)
Comment (jonasagx) [reply README.md]: The neighbor sections have H4 so I followed suit.

File: internal/formats/spdx22json/decoder_test.go (hunk 19 above)
Comment (jonasagx) [reply internal/formats/spdx22json/decoder_test.go]: it was necessary because it reads the file, to Dan's point I noticed this test doesn't help the encoder, so I removed it.

File: test/integration/encode_decode_cycle_test.go (hunk 22 above)
Comment (jonasagx) [reply test/integration/encode_decode_cycle_test.go]: outdated, please ignore

File: README.md (hunk 4 above)
Comment (luhring) [reply README.md:106]: I appreciate that, @jonasagx ! I actually think it's okay to leave out the explanation from the README, I just wanted to understand our strategy for myself.

I don't think we need to exclude the GH format necessarily, but I'm okay deferring this. If the error messages are good, like "destination format %q does not have an encoder" or something, I think users will know what to ask for if they want expanded support.

File: test/cli/power_user_cmd_test.go [hunk 23]
```diff
@@ -60,7 +60,7 @@ func TestPowerUserCmdFlags(t *testing.T) {
 			env: map[string]string{
 				"SYFT_SECRETS_REVEAL_VALUES": "true",
 			},
-			args: []string{"power-user", "docker-archive:" + secretsFixture},
+			args: []string{"power-user", "docker-archive:" + secretsFixture, "-vv"},
```
Comment (luhring) [inline test/cli/power_user_cmd_test.go:63]: Why these changes?

File: README.md [hunk 24]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
...
+
+```
+syft convert <ORIGINAL-SBOM-FILE> -o <NEW-SBOM-FORMAT>[=<NEW-SBOM-FILE>]
+```
+
+This feature is experimental and data might be lost when converting formats. Packages are the main SBOM component easily transferable across formats, while files and relationships are more likely to lose data. 
+
+The supported formats are:
+- Syft json
+- Syft table
```
Comment (kzantow) [inline README.md:108]: table is not supported for decoding, is it?

File: README.md [hunk 25]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
...
+- SPDX 2.2 json
+- SPDX 2.2 tag-value
+- cycloneDX 1.3 json
+- cycloneDX 1.3 xml
```
Comment (kzantow) [inline README.md:112]: Capitalization: CycloneDX, also we're at 1.4

File: README.md [hunk 26]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
...
+
+We support formats with wide community usage AND good encode/decode support by Syft. The recent GitHub format is niche and doesn't have a decoder, therefore was excluded.
```
Comment (kzantow) [inline README.md:114]: I wouldn't even mention GitHub format here

File: README.md [hunk 27]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
...
+
+The exception to the rule above is table format, which is the default output for two reasons:
+1. it matches syft's default SBOM output
+2. it is handy, as a summary, to print SBOMs as tables
+
+Conversion example:
+```sh
+syft alpine:latest -o json=sbom.syft.json # first let's generate a syft SBOM
+syft convert sbom.syft.json -o cyclonedx-json=sbom.cdx.json  # now let's convert it to cycloneDX
```
Comment (kzantow) [inline README.md:123]: Maybe just:
```suggestion
syft alpine:latest -o json=sbom.syft.json # generate a syft SBOM
syft convert sbom.syft.json -o cyclonedx-json=sbom.cdx.json  # convert it to cycloneDX
```

File: README.md [hunk 28]
```diff
@@ -92,6 +93,36 @@ The above output includes only software that is visible in the container (i.e.,
...
 
+#### Format conversion (experimental)
+
+The ability to convert existing SBOMs means you can create SBOMs in different formats quickly, without the need to regenerate the SBOM from scratch, which may take significantly more time.
+
+```
+syft convert <ORIGINAL-SBOM-FILE> -o <NEW-SBOM-FORMAT>[=<NEW-SBOM-FILE>]
+```
+
+This feature is experimental and data might be lost when converting formats. Packages are the main SBOM component easily transferable across formats, while files and relationships are more likely to lose data.
```
Comment (kzantow) [inline README.md:104]: > while files and relationships are more likely to lose data.
Maybe:
whereas files and relationships, as well as other information Syft doesn't support is more likely to be lost.

File: cmd/syft/cli/convert.go [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+
+const (
+	convertExample = `  {{.appName}} {{.command}} img.syft.json -o spdx-json                      convert a syft SBOM to spdx-json, output goes to stdout in table format, by default
+  {{.appName}} {{.command}} img.syft.json -o cyclonedx-json=img.cdx.json    convert a syft SBOM to CycloneDX, output goes to a file named img.cdx.json
+`
+)
+
+func Convert(v *viper.Viper, app *config.Application, ro *options.RootOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "convert sbom.json -o [FORMAT]",
```
Comment (kzantow) [inline cmd/syft/cli/convert.go:22]:
```suggestion
		Use:   "convert [SOURCE-SBOM] -o [FORMAT]",
```

File: cmd/syft/cli/convert/convert.go [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/anchore/syft/syft/sbom"
+)
+
+var ConvertibleFormats = []sbom.Format{
+	syftjson.Format(),
+	spdx22json.Format(),
+	spdx22tagvalue.Format(),
+	cyclonedxjson.Format(),
+	cyclonedxxml.Format(),
+	table.Format(),
```
Comment (kzantow) [inline cmd/syft/cli/convert/convert.go:27]: Why is table here?

File: README.md [hunk 31]
```diff
@@ -92,6 +93,31 @@ The above output includes only software that is visible in the container (i.e.,
...
+This feature is experimental and data might be lost when converting formats. Packages are the main SBOM component easily transferable across formats, whereas files and relationships, as well as other information Syft doesn't support, are more likely to be lost. 
+
+The supported formats are:
+- Syft json
+- SPDX 2.2 json
+- SPDX 2.2 tag-value
+- CycloneDX 1.4 json
+- CycloneDX 1.4 xml
+
+We support formats with wide community usage AND good encode/decode support by Syft. The recent GitHub format is niche and doesn't have a decoder, therefore was excluded.
```
Comment (kzantow) [inline README.md:113]: Still has:

> The recent GitHub format is niche and doesn't have a decoder, therefore was excluded.

File: README.md [hunk 32]
```diff
@@ -92,6 +93,31 @@ The above output includes only software that is visible in the container (i.e.,
...
+```
+
+This feature is experimental and data might be lost when converting formats. Packages are the main SBOM component easily transferable across formats, whereas files and relationships, as well as other information Syft doesn't support, are more likely to be lost. 
+
+The supported formats are:
+- Syft json
+- SPDX 2.2 json
+- SPDX 2.2 tag-value
+- CycloneDX 1.4 json
+- CycloneDX 1.4 xml
```
Comment (kzantow) [inline README.md:111]: One more nit here: capitalization of JSON and XML

File: cmd/syft/cli/convert/convert.go [hunk 33]
```diff
@@ -0,0 +1,41 @@
...
+	f, err := os.Open(userInput)
+	if err != nil {
+		return fmt.Errorf("failed to open SBOM file: %w", err)
+	}
+
+	sbom, _, err := syft.Decode(f)
+	if err != nil {
+		return fmt.Errorf("failed to decode SBOM: %w", err)
+	}
+	f.Close()
```
Comment (kzantow) [inline cmd/syft/cli/convert/convert.go:38]: should this be in a `defer` block?

File: cmd/syft/cli/options/writer.go [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+
+func IsSupportedFormat(format sbom.FormatID, supported []sbom.Format) bool {
+	for _, f := range supported {
+		if format == f.ID() {
+			return true
+		}
+	}
+
+	return false
+}
```
Comment (kzantow) [inline cmd/syft/cli/options/writer.go:21]: this is not needed anywhere

File: cmd/syft/cli/options/writer_test.go [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+		{
+			outputs: []string{"unknown"},
+			wantErr: func(t assert.TestingT, err error, bla ...interface{}) bool {
+				return assert.ErrorContains(t, err, "bad output format: 'unknown'")
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		_, err := MakeWriter(tt.outputs, "", tt.supported...)
```
Comment (kzantow) [inline cmd/syft/cli/options/writer_test.go:51]: signature changed, doesn't need `, tt.supported...`

File: internal/formats/common/spdxhelpers/to_syft_model.go [hunk 36]
```diff
@@ -42,6 +49,31 @@ func ToSyftModel(doc *spdx.Document2_2) (*sbom.SBOM, error) {
 	return s, nil
 }
 
+// NOTE(jonas): this is a hack, SPDX doesn't seem to support a way
```
Comment (kzantow) [inline internal/formats/common/spdxhelpers/to_syft_model.go:52]: Might not call this "a hack", just "this is our best effort to determine the scheme" or something

File: syft/formats_test.go [hunk 37]
```diff
@@ -44,7 +44,7 @@ func TestIdentify(t *testing.T) {
 }
 
 func TestFormats_EmptyInput(t *testing.T) {
-	for _, format := range formats {
+	for _, format := range AllFormats {
```
Comment (kzantow) [inline syft/formats_test.go:47]: revert this

**Review by kzantow — CHANGES_REQUESTED** [review]
Just a few more minor things + address the compilation failures

File: README.md (hunk 17 above)
Comment (kzantow) [reply README.md]: @spiffcs to answer your question: the `-o` hasn't changed, today it supports `-o <format>` and `-o <format>=<file>`, with multiple `-o` allowed. this still works with this `convert` command

File: cmd/syft/cli/convert/convert.go (hunk 33 above)
Comment (jonasagx) [reply cmd/syft/cli/convert/convert.go:38]: No, after decoding that file is no longer used and can be closed, defer would just postpone it to the end of the function.

File: README.md [hunk 38]
```diff
@@ -92,6 +93,29 @@ The above output includes only software that is visible in the container (i.e.,
...
+
+The ability to convert existing SBOMs means you can create SBOMs in different formats quickly, without the need to regenerate the SBOM from scratch, which may take significantly more time.
+
+```
+syft convert <ORIGINAL-SBOM-FILE> -o <NEW-SBOM-FORMAT>[=<NEW-SBOM-FILE>]
+```
+
+This feature is experimental and data might be lost when converting formats. Packages are the main SBOM component easily transferable across formats, whereas files and relationships, as well as other information Syft doesn't support, are more likely to be lost. 
+
+We support formats with wide community usage AND good encode/decode support by Syft. The supported formats are:
```
Comment (kzantow) [inline README.md:106]: Looks great 👍

File: cmd/syft/cli/convert/convert.go (hunk 33 above)
Comment (kzantow) [reply cmd/syft/cli/convert/convert.go:38]: I mean, there are 2 cases it will return before `f.Close()`, should it be in a `defer` block immediately following the `os.Open` line 29?

Comment (jonasagx) [reply cmd/syft/cli/convert/convert.go:38]: ops ops, you're totally right

File: internal/formats/common/spdxhelpers/to_syft_model.go (hunk 13 above)
Comment (kzantow) [reply internal/formats/common/spdxhelpers/to_syft_model.go:34]: We discussed this one, I think this is good to keep; it could be iterated on in the future to try to determine the source type in other ways

File: README.md [hunk 39]
```diff
@@ -92,6 +93,29 @@ The above output includes only software that is visible in the container (i.e.,
...
+- Syft JSON
+- SPDX 2.2 JSON
+- SPDX 2.2 tag-value
+- CycloneDX 1.4 JSON
+- CycloneDX 1.4 XML
+
+Conversion example:
+```sh
+syft alpine:latest -o json=sbom.syft.json # generate a syft SBOM
```
Comment (spiffcs) [inline README.md:115]: Were earlier comments force pushed away? I still had a question about this being listed as an alias for the root command and don't see the previous discussion on that here?

Comment (spiffcs) [inline README.md:115]: Should -o here be `json=sbom.syft.json`?

I thought valid values for this were `syft-json cyclonedx-xml cyclonedx-json github github-json spdx-tag-value spdx-json table text`

File: README.md [hunk 40]
```diff
@@ -92,6 +93,29 @@ The above output includes only software that is visible in the container (i.e.,
...
+syft convert sbom.syft.json -o cyclonedx-json=sbom.cdx.json  # convert it to CycloneDX
```
Comment (spiffcs) [inline README.md:116]: Same question here - Why is -o now shorthand for this `x=x` version rather than what the help output says are the valid options?

File: README.md (hunk 39 above)
Comment (kzantow) [reply README.md:115]: `json` is an alias for `syft-json`

Comment (kzantow) [reply README.md:115]: just like you can use `-o json` in syft...

File: README.md (hunk 40 above)
Comment (kzantow) [reply README.md:116]: it might be good to use the non-shorthand versions in the examples, good callout @spiffcs

**Review by kzantow — APPROVED** [review]
As far as I'm concerned, this is in great shape now. It might be a good idea not to use the `json` shorthand in the example, but I, personally, wouldn't hold up this PR for that detail. 👍

File: README.md (hunk 39 above)
Comment (spiffcs) [reply README.md]: So in this case the output file is `sbom.syft.json`? I understand `json` as the alias. What was confusing was that the command is outputting to a file called `sbom.syft.json`

Comment (spiffcs) [reply README.md]: No need to hole up the PR, but =<file> might be misleading without comments on where the output is going. Especially since we have other demos/examples where syft is just doing a `syft packages -o json alpine:latest > syft-sbom.json`

Comment (kzantow) [reply README.md]: FYI this is documented in the Syft readme: https://github.com/anchore/syft#multiple-outputs

Comment (kzantow) [reply README.md]: Having said that, I might also just use a "more normal" `-o <format>` either with `> <file>` or `--file <file>`

Comment (kzantow) [reply README.md]: This is a very minor issue to me I wouldn't hold up the PR for, though.
